### PR TITLE
Improve Hubble decoding performance for trace events

### PR DESCRIPTION
**Pull request by brancz**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

See each commit message for the detailed description, but these patches reduce the baseline CPU usage of Cilium Agents by 3-4%, by removing allocations. The decode functions themselves improved in runtime by 92-95%.

```release-note
Improve Hubble decoding performance for trace events
```

### CODE REVIEW

**Comment by chancez** [comment]
Verified sign-off manually and removed the label.

File: pkg/monitor/datapath_trace_test.go [hunk 1]
```diff
@@ -4,6 +4,10 @@
 package monitor
 
 import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/cilium/cilium/pkg/byteorder"
```
Comment (gandro) [inline pkg/monitor/datapath_trace_test.go:13]: CI is complaining about the import order here

File: pkg/monitor/datapath_trace_test.go [hunk 2]
```diff
@@ -49,3 +53,33 @@ func (s *MonitorSuite) TestDecodeTraceNotify(c *C) {
 	c.Assert(err, NotNil)
 
 }
+
+func (s *MonitorSuite) BenchmarkDecodeTraceNotifyVersion0(c *C) {
```
Comment (gandro) [inline pkg/monitor/datapath_trace_test.go]: Would you mind adding a few more unit tests here? Mostly to ensure that we notice if one changes the TraceNotify struct without also changing `decodeTraceNotifyVersion0`.

`TestDecodeTraceNotify` tests some of the basics, but I think we should have at least one test that has all fields non-zero and ensures that all fields are properly decoded.

**Review by gandro — COMMENTED** [review]
Amazing work, thank you! I've left two comments around the test file.

Side-note (not for this PR): We should probably do the same for other events (such as PolicyVerdict and DropNotify) too. We've already been doing it for `RecorderCapture`, but that one is rarely used.

File: pkg/monitor/datapath_trace.go [hunk 3]
```diff
@@ -43,6 +42,7 @@ const (
 )
 
 // TraceNotifyV0 is the common message format for versions 0 and 1.
+// This struct needs to be kept in sync with the function below.
```
Comment (kaworu) [inline pkg/monitor/datapath_trace.go]: I would prefer to refer to the function explicitely by name so that moving code around would not "break the comment".
```suggestion
// This struct needs to be kept in sync with the decodeTraceNotifyVersion0
// func.
```

File: pkg/monitor/datapath_trace.go [hunk 4]
```diff
@@ -60,13 +60,51 @@ type TraceNotifyV0 struct {
 	// data
 }
 
+// decodeTraceNotifyVersion0 decodes the trace notify message in 'data' into
+// the struct. This function needs to be kept in sync with the struct above.
```
Comment (kaworu) [inline pkg/monitor/datapath_trace.go]:
```suggestion
// the struct. This function needs to be kept in sync with the TraceNotifyV0
// struct.
```

File: pkg/monitor/datapath_trace.go [hunk 5]
```diff
@@ -60,13 +60,51 @@ type TraceNotifyV0 struct {
...
 }
 
+func (tn *TraceNotifyV1) decodeTraceNotifyVersion1(data []byte) error {
+	if err := tn.decodeTraceNotifyVersion0(data); err != nil {
+		return err
+	}
+
+	if len(data) < traceNotifyV1Len {
+		return errors.New("unexpected length of data to decode into TraceNotifyV1")
+	}
```
Comment (kaworu) [inline pkg/monitor/datapath_trace.go]: Could we move this guard before the call to `decodeTraceNotifyVersion0`? That way the receiver is not mutated on failure.

File: pkg/monitor/datapath_trace.go [hunk 6]
```diff
@@ -60,13 +60,51 @@ type TraceNotifyV0 struct {
...
+}
+
 // TraceNotifyV1 is the version 1 message format.
 type TraceNotifyV1 struct {
 	TraceNotifyV0
 	OrigIP types.IPv6
 	// data
 }
 
+func (tn *TraceNotifyV1) decodeTraceNotifyVersion1(data []byte) error {
```
Comment (kaworu) [inline pkg/monitor/datapath_trace.go:100]:
```suggestion
// decodeTraceNotifyVersion1 decodes the trace notify message in 'data' into
// the struct. This function needs to be kept in sync with the TraceNotifyV1
// struct.
func (tn *TraceNotifyV1) decodeTraceNotifyVersion1(data []byte) error {
```

File: pkg/monitor/datapath_trace.go [hunk 7]
```diff
@@ -60,13 +60,51 @@ type TraceNotifyV0 struct {
...
+	tn.DstLabel = identity.NumericIdentity(byteorder.Native.Uint32(data[20:24]))
+	tn.DstID = byteorder.Native.Uint16(data[24:26])
+	tn.Reason = data[26]
+	tn.Flags = data[27]
+	tn.Ifindex = byteorder.Native.Uint32(data[28:32])
+
+	return nil
+}
+
 // TraceNotifyV1 is the version 1 message format.
```
Comment (kaworu) [inline pkg/monitor/datapath_trace.go]: probably a good idea to have the same kind of comment as for v0 about keeping the `struct`/`func` in sync

```suggestion
// TraceNotifyV1 is the version 1 message format. This struct needs to be kept
// in sync with the decodeTraceNotifyVersion1 func.
```

**Review by kaworu — COMMENTED** [review]
Hi @brancz and thanks for the PR (and the stream!)

Some nitpicking comments but overall the patch LGTM.

File: pkg/monitor/datapath_trace_test.go (hunk 2 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace_test.go]: Yes we definitely need more coverage once we start "manually" decoding.

**Comment by brancz** [comment]
Thanks for the fixes on top, and sorry for not having gotten back to it yet! ❤️

**Comment by kaworu** [comment]
> Thanks for the fixes on top, and sorry for not having gotten back to it yet! heart

No problem @brancz, I'm adding a couple of tests and then the PR should be good to go once it passes CI.

File: pkg/monitor/datapath_trace_test.go (hunk 2 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace_test.go]: @gandro added a couple of tests (one for v1, one for v2) with non-zero value fields and should work regardless of the test runner's endianess, PTAL.

**Comment by kaworu** [comment]
Reworded the PR title and release note block into something more appropriate for release notes, leaving the details about the CPU cycle improvements in the commit messages.

File: pkg/monitor/datapath_trace_test.go [hunk 8]
```diff
@@ -4,48 +4,125 @@
...
+		uint16(0x19_1a),                          // DstID
+		uint8(0x1b),                              // Reason
+		uint8(0x1c),                              // Flags
+		uint32(0x1d_1e_1f_20),                    // Ifindex
 	}
 
-	err = DecodeTraceNotify(data, tn)
+	buf := bytes.NewBuffer(nil)
+	for _, v := range fields {
+		err := binary.Write(buf, byteorder.Native, v)
```
Comment (gandro) [inline pkg/monitor/datapath_trace_test.go]: We could probably should directly write a `TraceNotify` struct using `binary.Write` here, instead of assuming the order of fields in the test and writing them one by one. If the order of fields in `TraceNotify` were to change, this test would not catch it, I think.

You might even be able to use `testutils.MustCreateL3L4Payload` here.

**Review by gandro — COMMENTED** [review]
Awesome, looks good to me overall. I do think however we should make the unit tests a bit more robust, since I assume we want to use the same method for other *Notify structs as well going forward.

File: pkg/monitor/datapath_trace_test.go (hunk 8 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace_test.go]: Ah yes good catch, will update!

Comment (kaworu) [reply pkg/monitor/datapath_trace_test.go]: Updated as suggested, but couldn't use `testutils.MustCreateL3L4Payload` because of an import cycle, please take another look :pray:

**Review by gandro — APPROVED** [review]
Nice, thanks! Should we create a follow-up issue to give `{Drop,PolicyEvent,..}Notify` the same treatment?

File: pkg/monitor/datapath_trace.go [hunk 9]
```diff
@@ -60,6 +63,32 @@ type TraceNotifyV0 struct {
 	// data
 }
 
+// decodeTraceNotifyVersion0 decodes the trace notify message in 'data' into
+// the struct. This function needs to be kept in sync with the TraceNotifyV0
+// struct.
+func (tn *TraceNotifyV0) decodeTraceNotifyVersion0(data []byte) error {
+	// This eliminates the bounds check in the accesses to `data` below.
+	if len(data) < traceNotifyV0Len {
+		return errors.New("unexpected length of data to decode into TraceNotifyV0")
```
Comment (aspsk) [inline pkg/monitor/datapath_trace.go]: can we print this unexpected length here?

File: pkg/monitor/datapath_trace.go [hunk 10]
```diff
@@ -8,6 +8,7 @@ import (
 	"bytes"
```
Comment (aspsk) [inline pkg/monitor/datapath_trace.go]: Unrelated to code: commit message says "resulting in a ~92% reduction in runtime performance", 92% reduction is actually 12.5 slower :) did you mean `92% increase`?

Comment (aspsk) [reply pkg/monitor/datapath_trace.go]: Same comment about the next commit message

File: pkg/monitor/datapath_trace.go [hunk 11]
```diff
@@ -89,13 +87,30 @@ func (tn *TraceNotifyV0) decodeTraceNotifyVersion0(data []byte) error {
...
 	OrigIP types.IPv6
 	// data
 }
 
+// decodeTraceNotifyVersion1 decodes the trace notify message in 'data' into
+// the struct. This function needs to be kept in sync with the TraceNotifyV1
+// struct.
+func (tn *TraceNotifyV1) decodeTraceNotifyVersion1(data []byte) error {
+	if len(data) < traceNotifyV1Len {
+		return errors.New("unexpected length of data to decode into TraceNotifyV1")
```
Comment (aspsk) [inline pkg/monitor/datapath_trace.go]: same, please print the `len(data)` in the error, sometimes it helps to understand the reason of the misconfiguration

File: pkg/monitor/datapath_trace_test.go [hunk 12]
```diff
@@ -6,7 +6,6 @@ package monitor
 import (
```
Comment (aspsk) [inline pkg/monitor/datapath_trace_test.go:6]: Could you please add a brief summary on why we need to replace standard go benchmarks to gocheck to the commit message

File: pkg/monitor/datapath_trace_test.go [hunk 13]
```diff
@@ -123,40 +122,32 @@ func (s *MonitorSuite) TestDecodeTraceNotifyErrors(c *C) {
...
-func BenchmarkDecodeTraceNotifyVersion0(b *testing.B) {
+func (s *MonitorSuite) BenchmarkDecodeTraceNotifyVersion0(c *C) {
 	input := TraceNotifyV0{}
 	buf := bytes.NewBuffer(nil)
 
-	if err := binary.Write(buf, byteorder.Native, input); err != nil {
-		b.Fatal(err)
-	}
-
-	b.ReportAllocs()
```
Comment (aspsk) [inline pkg/monitor/datapath_trace_test.go]: is the call `b.ReportAllocs()` needs to be replaced by `c.ReportAllocs()` or it is not supported?

File: pkg/monitor/datapath_trace.go (hunk 10 above)
Comment (aspsk) [reply pkg/monitor/datapath_trace.go]: Another comment about the commit message is that \`\`\` may be omitted, and you can just shift the output block by 4 characters to the right. This will be read better in the text form, and github will display it as code as well

**Review by aspsk — CHANGES_REQUESTED** [review]
Looks great, thanks! I 've left some minor comments

File: pkg/monitor/datapath_trace_test.go (hunk 12 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace_test.go:6]: @brancz authored this commit so I'm not sure. I would guess because we already use gocheck for other tests through the file (although they are not benchmark). Seems like we're inconsistently using one or the other for benchmarks (e.g. [here](https://github.com/cilium/cilium/blob/c2b5e70535ef725b372b622be38b1467c6b7bd21/daemon/cmd/fqdn_test.go#L162) and [here](https://github.com/cilium/cilium/blob/c2b5e70535ef725b372b622be38b1467c6b7bd21/pkg/datapath/loader/loader_test.go#L246)).

File: pkg/monitor/datapath_trace_test.go (hunk 13 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace_test.go]: I've looked into this and as far as I can see the gocheck way is to [pass the` -check.b=true` flag](https://github.com/go-check/check/blob/10cb98267c6cb43ea9cd6793f29ff4089c306974/run.go#L42), which will [enable `BenchmarkMem` from the gocheck `RunConf`](https://github.com/go-check/check/blob/10cb98267c6cb43ea9cd6793f29ff4089c306974/run.go#L61). There is no `c.ReportAllocs()`.

File: pkg/monitor/datapath_trace.go (hunk 9 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace.go]: done

File: pkg/monitor/datapath_trace.go (hunk 10 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace.go]: fixed

File: pkg/monitor/datapath_trace.go (hunk 11 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace.go]: done

File: pkg/monitor/datapath_trace_test.go (hunk 13 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace_test.go]: I wasn't able to run the gocheck benchmark though, https://stackoverflow.com/questions/24421535/how-do-you-use-the-benchmark-flags-for-the-go-golang-gocheck-testing-framework is the only thing that I found but wasn't helpful. I'm tempted to drop the last commit.

**Review by aspsk — APPROVED** [review]

File: pkg/monitor/datapath_trace_test.go (hunk 13 above)
Comment (kaworu) [reply pkg/monitor/datapath_trace_test.go]: Dropped the last commit.
